### DISCUSSION TITLE
[Question] Ocaml's idiom for repetitive task

### DISCUSSION THREAD

**Discussion by jidnjs** [prosyslab-classroom/cs424-program-reasoning#441] [Q&A] [answered]
### Your Name

Jiwon Kang (강지원)

### Question

반복적으로 수행하다가 어느 순간에 종료해야 하는 프로그램을 Ocaml로 작성한다고 할 때, 
주로 3가지 방법이 있을 것 같은데 Ocaml은 주로 어떤 방법을 지향하는지 궁금하여 질문 드립니다.
1. 재귀 함수 사용
- 꼬리 호출 최적화가 가능하다면 가장 알맞은 방법이라고 생각하여 고려하게 되었습니다.
```
let rec do_task_rec () = 
  (* snip *)
  if flag then () else do_task_rec ()
```
2. 가변 플래그 + while 문에서의 플래그 확인
- 첫 번째와 달리 재귀 함수로 만드는 추가적인 과정이 불필요하다는 점에서 고려하게 되었습니다.
```
let do_task () =
  let flag = ref false in
  while !flag do
    (* snip *)
  done
```
3. try ... with 구문 내의 무한 루프 + raise exception 구문을 사용하여 탈출
- 두 번째 방법에서 task가 짧지만 많이 반복하기에, flag를 역참조하는 것이 성능에 영향을 끼칠 수 있다고 생각해 고려하게 되었습니다.
```
exception TaskDone
let do_task () =
  try
    while true do
      (* snip *)
    done
  with
  | TaskDone -> ()
  | e -> raise e
```
---
When writing a program in OCaml that performs a repetitive task and needs to terminate at a certain point, I considered three possible approaches and wondered which approach OCaml typically recommends. Here's the question:
1. Using a recursive function
- I considered this method as it seems most appropriate if tail-call optimization is possible.
```
let rec do_task_rec () = 
  (* snip *)
  if flag then () else do_task_rec ()
```
2. Using a mutable flag and checking it within a while loop
- Unlike first method, this avoids the additional process of creating a recursive function, which led me to consider it.
```
let do_task () =
  let flag = ref false in
  while !flag do
    (* snip *)
  done
```
3. Using a try ... with block with an infinite loop and raising an exception to exit
- This method came to mind because if the task is short but needs to repeat many times, the overhead of dereferencing the flag in second method might impact performance.
```
exception TaskDone
let do_task () =
  try
    while true do
      (* snip *)
    done
  with
  | TaskDone -> ()
  | e -> raise e
```

**Comment by spearo2** [comment, marked as answer]
함수형 언어에서는 반복문의 사용을 지향하지 않습니다.
Ocaml은 `while`과 같은 반복문을 지원하지만, 이를 구현에 사용하는 것을 권장하지 않습니다.
대신에 fold, map, iter과 같은 함수를 사용하거나 말씀하신 것과 같은 꼬리 호출을 사용하여 동일한 기능을 구현할 수 있습니다.
이는 반복문이 소위 ＂functional＂하지 않기 때문입니다.

다음의 링크가 도움이 되실 것 같습니다.
[how-to-code-with-no-loops](https://sarakhandaker.medium.com/how-to-code-with-no-loops-8ed815624aae)

---

Functional programming languages discourage the use of loops.
While OCaml does support loops like `while`, it is not recommended to use them in implementations.
Instead, you can achieve the same functionality using functions like fold, map, and iter, or by employing tail recursion as you mentioned.
This is because loops are not considered "functional" in nature.

The following link might be helpful:
[how-to-code-with-no-loops](https://sarakhandaker.medium.com/how-to-code-with-no-loops-8ed815624aae)

**Comment by KihongHeo** [comment]
좋은 질문이네요. 저는 1을 추천합니다. 간단하고 아름답기 때문이지요.

2번은 좋지 않습니다. 가변 플래그가 많아지면 코드의 올바름을 판단하기 어렵습니다. 전체를 부분의 합으로 이해하기가 어려워서 그렇습니다. 수업시간에 다룬 프로그램 검증을 생각해보세요. 가변 플래그, 전역 변수 등이 있을때 얼마나 검증이 복잡해질까요?

3번도 비슷한 이유로 지양하는게 좋습니다.

---

That’s a great question. I recommend option 1 because it is simple and elegant.

Option 2 is not ideal. When there are too many mutable flags, it becomes difficult to determine the correctness of the code. This is because it’s harder to understand the whole as a sum of its parts. Think about program verification, which we discussed in class. How much more complicated does verification become when mutable flags or global variables are involved?

Option 3 should also be avoided for similar reasons.
